### DISCUSSION TITLE
ABB upload Modules into existing Programm

### DISCUSSION THREAD

**Discussion by Neunzehnneunzig** [visose/Robots#245] [Q&A] [answered]
Hi everybody! 
Has anyone found a strategy to upload Modules directly into an existing program on an IRC5 Controller with the visose grasshopper uploader without deleting the existing program?

I don't want to delete existing Modules on the controller and I would prefer not to make them all system modules. I also don't want to go the Robot Studio route because I want to directly generate a multitude of modules and procedures. 

I've seen in the log that the Plugin deletes the existing Program before uploading an altered version, even if they share the same name. 

Has someone a idea how to make it behave the same way if I would load a programm with the same name in Robot Studio. So that it only adds missing modules or overwrites altered ones, but does not delete ones that are not in the new program version. 

Thanks for your help!

**Comment by visose** [comment, marked as answer]
Deleting the folder in the controller instead of just replacing files is done on purpose. You could change the program in a way that creates less files, so some files would remain unused and possibliy conflict with the new changes.
Why do you need the extra modules for? If it's not much code, you can include them in Grasshopper as part of the modules the plugin generates using the the "custom code" component.

**Reply by Neunzehnneunzig** [reply]
Hello @visose,
thank you for your answer!
I want to have separate modules, so that I can seamlessly switch between grasshopper and robot studio.
I've got a robot that has to fullfil multiple tasks. Some are better to realize with grasshopper and some are better done in robot studio.
I realy like the instant upload function for complex pathplanning and 3d printing applications. In short applications that require intensive changes in the programm. 

Right now I have integrated the code that your plugin generates in a way that requires copy pasting it into robot studio. Thus always replacing the former version of the code. (e.g ModuleN with procedures n1 to n5 gets replaced by ModuleN with procedures n1 to n70).

That is working without problems. But this manual step limits the flexibilty and speed that liveuploading through your plugin offers and forces it to stay beyond it's potential.

I understand the problem you are describing and that changing the auto delete feature might create problems for some users. 
So this definetly would not make sense as a default.

But beeing able to switch between the default (delete and create) and an alternative (just create or if existing replace) would be amazing. This would offer the possibilty to stream different programms through a loop while still being able to do all the nice and convenient things in robot studio without compromise.

That would just be amazing.

**Reply by visose** [reply]
If you're already using robotstudio, you can also try using the SaveProgram component to save to disk and use robotstudio to upload, might be faster if it's multiple files you're copy pasting.

**Reply by Neunzehnneunzig** [reply]
As soon as i start copy pasting I loose all dynamic interaction with the code.
That just has been a temporary makeshift workaround for me to figure out if it would work at all like i imagined.
This dynamic flexibility is what I'm after.

I totally understand that this is not a priority or something that would make sense for eveybody.
But is it theoretically possible to achieve something like I described?
Maybe even with a reasonable amount of work?
Or is it just not possible from the ABB side of things?

**Reply by visose** [reply]
I'm not sure what you mean with dynamic interaction, but I meant to save them to disk and upload them from RobotStudio so that you don't have to copy paste.
Like this:
<img width="955" alt="image" src="https://github.com/visose/Robots/assets/5129614/0f160f67-cb3f-4900-b743-e2b92ff3d543">
